### PR TITLE
feat(eslint-plugin): no-empty-interface - add allowSingleExtend option

### PR DESCRIPTION
**Pull request by timkraut**
Hi,

this PR adds an option to allow a single extend (option: `allowSingleExtend: true|false`) to the rule `no-empty-interface`. Our use case for this is the following:

### Use case
We are building a React application with TypeScript. In React, you typically have components which might look like this:
```ts
const ProductTile<ProductTileProps> = () => {
  // Component implementation
}
```

#### Problem
Let's say our `<ProductTile>` component accepts the exact same props as the members defined in an `interface Product { ... }`. We could change the definition of the `<ProductTile>` component as follows to support that:
```ts
const ProductTile<Product> = () => {
  // Component implementation
}
```

But once we have to add an additional prop (e.g. `onClick?(event: MouseEvent<HTMLButtonElement>): void`), we have the issue that we would have to extend `Product` (which doesn't make sense as `onClick` doesn't belong to a `Product`) or replace the interface.

#### Solution
Instead, we would prefer to define the component as follows initially:
```ts
interface ProductTileProps extends Product {}

const ProductTile<ProductTileProps> = () => {
  // Component implementation
}
```

Now if we add an additional prop like `onClick`, we could simply add it to the interface `ProductTileProps` and we're done:
```ts
const ProductTileProps extends Product {
  onClick?(event: MouseEvent<HTMLButtonElement>): void
}

// Component stays 100% the same
```

This PR allows to do that while still warning if empty interfaces like `interface Foo {}` are used. It's an optional configuration property so that the default behavior is not changed.

P.S.: This is my first PR to change an ESLint-style rule. Let me know if there is anything I should do differently. Thanks a lot!

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/no-empty-interface.md [hunk 1]
```diff
@@ -38,6 +38,12 @@ interface Bar {
 interface Baz extends Foo, Bar {}
 ```
 
+### Options
+
```
Comment (armano2) [inline packages/eslint-plugin/docs/rules/no-empty-interface.md]: can you add json example of default config for rule:

```json
{
  "rule-name": ["error", {
    "allowSingleExtend": false
  }]
}

File: packages/eslint-plugin/lib/rules/no-empty-interface.js [hunk 2]
```diff
@@ -20,32 +26,61 @@ module.exports = {
...
+          allowSingleExtend: {
+            type: 'boolean'
+          }
+        }
+      }
+    ],
+    messages: {
+      empty: 'An empty interface is equivalent to `{}`.',
+      extendsWithoutMembers:
+        'An interface declaring no members is equivalent to its supertype.'
```
Comment (armano2) [inline packages/eslint-plugin/lib/rules/no-empty-interface.js]: thank you for updating rule to use `messages`, can you update tests to

Comment (timkraut) [reply packages/eslint-plugin/lib/rules/no-empty-interface.js]: Should be done

File: packages/eslint-plugin/docs/rules/no-empty-interface.md (hunk 1 above)
Comment (timkraut) [reply packages/eslint-plugin/docs/rules/no-empty-interface.md]: Should be done

File: packages/eslint-plugin/tests/lib/rules/no-empty-interface.js [hunk 3]
```diff
@@ -37,14 +37,24 @@ interface Bar {
...
+      options: [{ allowSingleExtend: true }]
+    }
   ],
   invalid: [
     {
       code: 'interface Foo {}',
       errors: [
         {
-          message: 'An empty interface is equivalent to `{}`.',
+          message: rule.meta.messages.empty,
```
Comment (armano2) [inline packages/eslint-plugin/tests/lib/rules/no-empty-interface.js]: `messageId: 'empty'`

File: packages/eslint-plugin/tests/lib/rules/no-empty-interface.js [hunk 4]
```diff
@@ -67,11 +77,11 @@ interface Foo {
...
 
 interface Bar extends Foo {}
-            `,
+        `,
+      options: [{ allowSingleExtend: false }],
       errors: [
         {
-          message:
-            'An interface declaring no members is equivalent to its supertype.',
+          message: rule.meta.messages.extendsWithoutMembers,
```
[the remark this answers is not in the document]
Comment (timkraut) [reply packages/eslint-plugin/tests/lib/rules/no-empty-interface.js]: Should be done

File: packages/eslint-plugin/tests/lib/rules/no-empty-interface.js [hunk 5]
```diff
@@ -54,7 +64,7 @@ interface Baz extends Foo, Bar {}
       code: 'interface Foo extends {}',
       errors: [
         {
-          message: 'An empty interface is equivalent to `{}`.',
+          message: rule.meta.messages.empty,
```
[the remark this answers is not in the document]
Comment (timkraut) [reply packages/eslint-plugin/tests/lib/rules/no-empty-interface.js]: Should be done

File: packages/eslint-plugin/tests/lib/rules/no-empty-interface.js (hunk 3 above)
Comment (timkraut) [reply packages/eslint-plugin/tests/lib/rules/no-empty-interface.js]: Should be done

File: packages/eslint-plugin/docs/rules/no-empty-interface.md [hunk 6]
```diff
@@ -38,6 +38,23 @@ interface Bar {
...
+This rule accepts a single object option with the following default configuration:
+
+```json
+{
+  "@typescript-eslint/no-empty-interface": [
+    "error",
+    {
+      "allowSingleExtend": false
```
Comment (JamesHenry) [inline packages/eslint-plugin/docs/rules/no-empty-interface.md]: Sorry to be that guy :D, but I found it jarring that it was `...Extend`, singular.

IMO it should either be `allowSingleExtends` or `allowSingleExtension`, probably the first one to better match the keyword and AST.

Comment (timkraut) [reply packages/eslint-plugin/docs/rules/no-empty-interface.md]: Don't worry! Better we do the right thing before having a bad API afterwards. I had been thinking quite some time about the name and I wasn't 100% sure if that's a good one so I'm totally open for a change here. I will update the PR to use `allowSingleExtends`

Comment (timkraut) [reply packages/eslint-plugin/docs/rules/no-empty-interface.md]: I've updated the PR. Please check again

**Comment by bradzacher** [comment]
Hey @timkraut, sorry for the pain, but we just merged a big PR to convert the plugin to typescript.
Please have a look through the source code and migrate your PR to typescript as well.
Let me know if you've got any questions!

**Comment by timkraut** [comment]
@bradzacher Don't worry, it totally makes sense to use TypeScript in this case. I've converted the PR to TypeScript. Please check again

**Pull request merged**